Add document expiry support

mrd1234 opened the pull request:
- Added orleans configuration section with collection of grain expiry values.
- Updated WriteAsync method to check for and use expiry value where available for each grain type.
- Added unit test for reading orleans config section.

mrd1234 committed:
Added orleans configuration section with collection of grain expiry values.
Updated WriteAsync method to check for and use expiry value where available for each grain type.
Added unit test for reading orleans config section.

mrd1234 committed:
Fixed code review issues.

mrd1234 committed:
Merge pull request #1 from mrd1234/feature/document-expiry-support

Feature/document expiry support

ashkan-saeedi-mazdeh commented:
LGTM
@MikeGoldsmith  Can you take a look?

MikeGoldsmith commented on CouchBaseStorage/Configuration/OrleansConfigurationExtensions.cs:
```diff
@@ -0,0 +1,31 @@
+    /// </summary>
+    public static class OrleansConfigurationExtensions
+    {
+        public static Dictionary<string, TimeSpan> ReadDocumentExpiryConfiguration()
```
I think `GetGrainExpiries` would be clearer here.

MikeGoldsmith commented on CouchBaseStorage/Configuration/OrleansConfigurationExtensions.cs:
```diff
@@ -0,0 +1,31 @@
+
+            foreach (GrainExpiry documentExpiry in grainExpiriesConfig.GrainExpiries)
+            {
+                var expiry = TimeSpan.Parse(documentExpiry.Expiry);
```
I think we should use `TimeSpan.TryParse` or wrap in try/catch blocks to prevent an exception bubbling out if parsing bad / invalid data.

mrd1234 replied:
I'm a bit conflicted on this one. Originally I was using tryparse and defaulting to TimeSpan.Zero if it failed to parse, but then I thought it might be better to let the exception bubble up so it is clear that the expiry configuration is incorrect and that documents will not have an expiry set.

I think throwing the exception is pretty standard behaviour, for instance when a string is provided where an integer is expected. I could create a custom exception that makes it more clear what the issue is if you prefer.

Another approach is to add an app setting that lets the user specify whether they want an exception thrown or to set no expiry for any grain type where expiry value is invalid.

Let me know what you want me to do and I'll implement it. For now I have left it.

MikeGoldsmith replied:
Yeah, I understand your concern here and agree blindly using a default value is also not great behaviour. Lets go with a custom exception to make sure the error is clear to the end user.

MikeGoldsmith commented on CouchBaseStorage/Configuration/OrleansDocumentExpiry/OrleansConfigurationSection.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace CouchBaseProviders.Configuration.OrleansDocumentExpiry
+{
+    public class OrleansConfigurationSection : ConfigurationSection
```
I'm worried using `OrleansConfigurationSection` as it may clash with other implementations (eg official Orleans). How about something like `CouchbaseOrleansConfigurationSection`?

MikeGoldsmith commented on CouchBaseStorage/Configuration/OrleansDocumentExpiry/OrleansGrainExpiryConstants.cs:
```diff
@@ -0,0 +1,16 @@
+
+        public const string GrainTypePropertyName = "grainType";
+
+        public const string ExpiryPropertyName = "timespan";
```
I think the property name of `expiresAt` or `expiresIn` is more accurate than `timespan`.

mrd1234 replied:
I originally named it expiry but I thought that timespan made it more clear the type of data expected. I've renamed it to expiresIn now.

The readme file should probably have a section relating to expiry configuration to make this clear to consumers of this provider.

MikeGoldsmith commented on CouchBaseStorage/Configuration/OrleansDocumentExpiry/OrleansGrainExpiryConstants.cs:
```diff
@@ -0,0 +1,16 @@
+        /// <summary>
+        /// Path to target <see cref="OrleansConfigurationSection"/> configuration XML element.
+        /// </summary>
+        public static readonly string OrleansConfigurationSectionPath = "orleans";
```
Same as with `OrleansConfigurationSection`, this has the opportunity to clash with other implementations. Also, as you can name them whatever you want and/or nest configsections into groups, maybe we could pass in the path to the Read method?

MikeGoldsmith commented on CouchBaseStorage/CouchBaseStorageProvider.cs:
```diff
@@ -61,12 +64,17 @@ public class CouchBaseDataManager : IJSONStateDataManager
         /// <param name="bucketName">Name of the bucket that this manager should operate on.</param>
         /// <param name="clientConfig">Configuration object for the database client</param>
-        public CouchBaseDataManager(string bucketName, Couchbase.Configuration.Client.ClientConfiguration clientConfig)
+        public CouchBaseDataManager(string bucketName, Couchbase.Configuration.Client.ClientConfiguration clientConfig, Dictionary<string, TimeSpan> documentExpiries)
```
If we added a new constructor overload that took the document expiries it would mean we didn't have to change `CouchBaseMembershipProvider` or `CouchBaseDataManagerTests`.

MikeGoldsmith commented on CouchBaseStorage/CouchBaseStorageProvider.cs:
```diff
@@ -131,7 +141,7 @@ public CouchBaseDataManager(string bucketName, Couchbase.Configuration.Client.Cl
             throw result.Exception;
         }
-
+        
```
Whitespace

MikeGoldsmith commented on CouchBaseStorage/CouchBaseStorageProvider.cs:
```diff
@@ -61,12 +64,17 @@ public class CouchBaseDataManager : IJSONStateDataManager
         protected IBucket bucket;
 
+        /// <summary>
+        /// 
```
Empty summary. The code documentation is not complete so I'd be happy to either add something or remove the summary.

MikeGoldsmith reviewed (changes requested):
@mrd1234 I think what you've provided here is really good. I'm a little concerned about name clashes and have a couple of other suggestions / comments.
Excellent first draft.

mrd1234 committed:
Changed namespaces and class names to avoid name conflicts.
Renamed method used to load grain expiries,
Changed name of confgration expiry attribute.
Added unit test.

mrd1234 commented:
I think I have addressed everything you mentioned except for the exception handling (see comment).

Let me know if I have missed anything or you've thought of anything else you want me to look at.

mrd1234 committed:
Added checks to prevent duplicate grain type expiry values causing exception plus expiry values with empty grain type.

mrd1234 committed:
Renamed file to match class name.

mrd1234 committed:
Added custom exception for invalid timespan formats.
Modified config processnig to identify invalid timestamps, wrap them in an aggregateexception and throw it.
Renamed test file for consistency,

mrd1234 committed:
Changed inline out variable declaration to fix build error.

mrd1234 committed:
Update README.md

mrd1234 committed:
Update README.md

mrd1234 committed:
Added document expiry section to readme.

mrd1234 committed:
Updated readme.

mrd1234 commented:
I updated the readme file with document expiry details but looks like appveyor doesn't like it.

mrd1234 committed:
Merge branch 'master' into feature/document-expiry-support

mrd1234 committed:
Updated exception text.

mrd1234 committed:
Update README.md

mrd1234 committed:
Update README.md

mrd1234 committed:
Added custom aggregate exception to replace generic version.

Commit:
Simplified exception handling.

mrd1234 commented:
Hi guys,

Sorry for so many commits, but I think it's all sorted now.

I updated the readme file, and added custom exception handler. I did also create an aggregate exception to handle oth invalid expiry values and duplicate document expiry values per grain type, but then realised the duplicates seem to be ignored and not loaded by the config system anyway so I simplified it back to a single exception type.

Cheers

MikeGoldsmith commented:
Thanks @mrd1234 - I'll take a look soon.

MikeGoldsmith commented:
@mrd1234 looks good, thanks for the updates
@ashkan-saeedi-mazdeh you can merge this now

mrd1234 commented:
Thanks @MikeGoldsmith
